### PR TITLE
Register/track all the advancements properly.

### PR DESCRIPTION
**Pull request by Cybermaxke**
Better catch them all before they slip into more exceptions.

### CODE REVIEW

File: src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancement.java [hunk 1]
```diff
@@ -93,19 +93,19 @@
...
-        return SpongeImplHooks.isMainThread() || AdvancementRegistryModule.getInstance().isInInitialRegistration();
+        return SpongeImplHooks.isMainThread();
     }
 
     @SuppressWarnings("ConstantConditions")
     @Inject(method = "<init>", at = @At("RETURN"))
     private void onInit(ResourceLocation id, @Nullable Advancement parentIn, @Nullable DisplayInfo displayIn,
             AdvancementRewards rewardsIn, Map<String, Criterion> criteriaIn, String[][] requirementsIn, CallbackInfo ci) {
         // Don't do anything on the client, unless we're performing registry initialization
-        if (!this.isMainThread()) {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancement.java]: Why did you remove the this.?

File: src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java [hunk 2]
```diff
@@ -58,8 +60,8 @@
 
     @Shadow @Final private static Logger LOGGER;
 
-    @Shadow @Final private Map<ResourceLocation, Advancement> advancements;
-    @Shadow @Final private Set<Advancement> roots;
+    @Shadow @Final private Map<ResourceLocation, Advancement> advancements = new AdvancementMap();
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java]: This literally makes no sense. You're shadowing a field that is finalized in the actual class then initializing it here. Can't do that.

File: src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java [hunk 3]
```diff
@@ -58,8 +60,8 @@
...
+    @Shadow @Final private final Set<Advancement> roots = new RootAdvancementSet();
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java]: Same as above.

File: src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java (hunk 2 above)
Comment (Zidane) [reply src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java]: As well as you'll run into mod compatibility issues with replacing the map reference that Vanilla is using..

Comment (Cybermaxke) [reply src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java]: It's working fine like this. But I can move it if you want.

Comment (Cybermaxke) [reply src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java]: Same is done in: https://github.com/SpongePowered/SpongeCommon/blob/df3266d7b2923018f111df3a4db442e6a4882841/src/main/java/org/spongepowered/common/mixin/core/world/gen/MixinChunkProviderServer.java#L90

Comment (Cybermaxke) [reply src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java]: `@Mutable` is only used for setting fields after it's initialized

File: src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancement.java [hunk 4]
```diff
@@ -93,19 +93,19 @@
     @Nullable private Advancement tempParent;
 
     private void checkServer() {
-        checkState(this.isMainThread());
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancement.java]: Formatting change? I'd rather keep `this.`

File: src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java (hunk 2 above)
Comment (gabizou) [reply src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java]: The only reason why we do that for the chunk provider server is to provide better caching of the previously requested chunk, reducing lookups. If it's doing the same with the `AdvancementMap`, you should name it more appropriately.

File: src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java [hunk 5]
```diff
@@ -70,18 +72,13 @@ private void onLoadAdvancements(Map<ResourceLocation, Advancement.Builder> advan
         if (!SpongeImplHooks.isMainThread() || !ShouldFire.GAME_REGISTRY_EVENT_REGISTER) {
             return;
         }
-
+        final SpongeGameRegistryRegisterEvent<org.spongepowered.api.advancement.Advancement> event =
+                new SpongeGameRegistryRegisterEvent<>(Cause.of(EventContext.empty(), SpongeImpl.getRegistry()),
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java]: Why are you creating the cause yourself and not delegating to the cause stack manager?

Comment (gabizou) [reply src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java]: Which, in this case, you'd still want to put it in the try statement so the ListenerPhaseContext has a chance to provide some additional context

File: src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java (hunk 2 above)
Comment (Cybermaxke) [reply src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java]: I am just making sure that all the Advancements are getting captured and registered using the map.

File: src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java (hunk 5 above)
Comment (Cybermaxke) [reply src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java]: It's a mistake, so I will fix it

File: src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java [hunk 6]
```diff
@@ -58,8 +58,8 @@
 
     @Shadow @Final private static Logger LOGGER;
 
-    @Shadow @Final private Map<ResourceLocation, Advancement> advancements;
-    @Shadow @Final private Set<Advancement> roots;
+    @Shadow @Final private final Map<ResourceLocation, Advancement> advancements = new AdvancementMap();
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancementList.java]: Double check what the AP output is when compiling this mixin, I'm not sure whether `@Mutable` would be required. Pinging @Aaron1011 @Mumfrey for clarifying input?

File: src/main/java/org/spongepowered/common/registry/type/advancement/AdvancementMap.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+ */
+package org.spongepowered.common.registry.type.advancement;
+
+import net.minecraft.advancements.Advancement;
+import net.minecraft.util.ResourceLocation;
+import org.spongepowered.common.SpongeImplHooks;
+
+import java.util.HashMap;
+
+public final class AdvancementMap extends HashMap<ResourceLocation, Advancement> {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/registry/type/advancement/AdvancementMap.java:33]: Can you javadoc what this class is doing and why and how we're using it?

**Comment by bloodmc** [comment]
@Cybermaxke @gabizou is this PR ready?

**Comment by Cybermaxke** [comment]
@bloodmc Don't forget to merge this before RB ;)

**Comment by XakepSDK** [comment]
Advancement loading errors
https://gist.github.com/XakepSDK/249e9ca265f5c6b5c1ccb069ea6d4aa2
```
Mods (50): Minecraft, Minecraft Coder Pack, Forge Mod Loader, Minecraft Forge, SpongeAPI, SpongeForge, AI Improvements, AbyssalCraft, AbyssalCraft Integration, AppleCore, AppleSkin, Applied Energistics 2, BetterBedrockGen, Biomes O' Plenty, CoFH Core, CoFH World, CodeChicken Lib, CosmeticArmorReworked, Diet Hopper, FoamFix, FoamFixCore, InstantUnify, Inventory Tweaks, JourneyMap, Just Enough Items, Just Enough Pattern Banners, LagGoggles, LegibleNei, Mantle, Not Enough Items, NuclearCraft, Pam's Cookables, Pam's HarvestCraft, Quick Leaf Decay, Reborn Core, Redstone Flux, Roguelike Dungeons, Simply Jetpacks 2, Tech Reborn, The Betweenlands, Thermal Dynamics, Thermal Expansion, Thermal Foundation, Tinker's JEI, Tinkers' Construct, UniDict, WAIM, Waila, WanionLib, What Are We Looking At

Plugins (10): Minecraft, Minecraft Coder Pack, SpongeAPI, SpongeForge, GriefPrevention, LuckPerms, MagiBridge, Nucleus, Total Economy, UltimateChat

SpongeForge
    Minecraft: 1.12.2
    SpongeAPI: 7.1.0-SNAPSHOT-51caddf
    SpongeForge: 1.12.2-2705-7.1.0-BETA-3355
Minecraft Forge: 14.23.4.2739
```
